sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,895 @@
+        path to the caching directory.
+
+    n_jobs : int, default=4
+        Number of parallel jobs to run in core distance computations (if
+        supported by the specific algorithm). For `n_jobs<0`,
+        `(n_cpus + n_jobs + 1)` are used.
+
+    cluster_selection_method : str, default='eom'
+        The method used to select clusters from the condensed tree. The
+        standard approach for HDBSCAN* is to use an Excess of Mass algorithm
+        to find the most persistent clusters. Alternatively you can instead
+        select the clusters at the leaves of the tree -- this provides the
+        most fine grained and homogeneous clusters. Options are:
+        - `eom`
+        - `leaf`
+
+    allow_single_cluster : bool, default=False
+        By default HDBSCAN* will not produce a single cluster. Setting this to
+        `True` will allow single cluster results in the case that you feel this
+        is a valid result for your dataset.
+
+    metric_params : dict, default=None
+        Arguments passed to the distance metric.
+
+    Returns
+    -------
+    labels : ndarray, shape (n_samples, )
+        Cluster labels for each point.  Noisy samples are given the label -1.
+
+    probabilities : ndarray, shape (n_samples, )
+        Cluster membership strengths for each point. Noisy samples are assigned
+        0.
+
+    single_linkage_tree : ndarray, shape (n_samples - 1, 4)
+        The single linkage tree produced during clustering in scipy
+        hierarchical clustering format
+        (see http://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html).
+
+    References
+    ----------
+
+    .. [1] Campello, R. J., Moulavi, D., & Sander, J. (2013, April).
+       Density-based clustering based on hierarchical density estimates.
+       In Pacific-Asia Conference on Knowledge Discovery and Data Mining
+       (pp. 160-172). Springer Berlin Heidelberg.
+
+    .. [2] Chaudhuri, K., & Dasgupta, S. (2010). Rates of convergence for the
+       cluster tree. In Advances in Neural Information Processing Systems
+       (pp. 343-351).
+
+    .. [3] Malzer, C., & Baum, M. (2019). A Hybrid Approach To Hierarchical
+       Density-based Cluster Selection. arxiv preprint 1911.02282.
+    """
+    if min_samples is None:
+        min_samples = min_cluster_size
+
+    # Checks input and converts to an nd-array where possible
+    if metric != "precomputed" or issparse(X):
+        X = check_array(
+            X, accept_sparse="csr", force_all_finite=False, dtype=np.float64
+        )
+    elif isinstance(X, np.ndarray):
+        # Only non-sparse, precomputed distance matrices are handled here
+        # and thereby allowed to contain numpy.inf for missing distances
+
+        # Perform check_array(X) after removing infinite values (numpy.inf)
+        # from the given distance matrix.
+        tmp = X.copy()
+        tmp[np.isinf(tmp)] = 1
+        check_array(tmp)
```

Comment:
I think this code is indirectly checking `np.isnan(X).any()`. Can we check this directly?

Also does this mean that `nan` is not accepted when `metric=="precomputed"`?

Reply:
It also needs to do the rest of the usual input validation, e.g. `ensure_2d` so I think it's fine as-is.

Correct. `np.nan` is only valid as a value in a dense input array. The analog for `"precomputed"` would be `np.inf` since "missing data" denoted by `np.nan` in the input get handled as points with "infinite distance", hence `np.inf` in a distance array.

Reply:
Assuming `metric="precomputed"`, if there is an `inf` in the dataset, then `HDBSCAN.fit` errors before calling `hdbscan`:

https://github.com/Micky774/scikit-learn/blob/ed6d17dcb15b54feafa26f974e16c010ad3b587c/sklearn/cluster/_hdbscan/hdbscan.py#L728-L730

Am I missing something?

Reply:
Ah yes, the proper handling was included in `hdbscan` the function, but not the estimator. This is resolved after the refactor and removal of `hdbscan`.